piecedirectory/piecedirectory.go

```diff
@@ -347,61 +349,49 @@ type cachedSectionReader struct {
 	types.SectionReader
 	ps       *PieceDirectory
 	pieceCid cid.Cid
-	// The number of references to the piece reader - only accessed under prCacheLk
-	refs int
 	// Signals when the underlying piece reader is ready
 	ready chan struct{}
 	// err is non-nil if there's an error getting the underlying piece reader
 	err error
 }
 
 func (r *cachedSectionReader) Close() error {
-	r.ps.prCacheLk.Lock()
-	defer r.ps.prCacheLk.Unlock()
-
-	r.refs--
-	if r.refs == 0 {
-		r.ps.pieceReaderCache.Remove(r.pieceCid)
-		return r.SectionReader.Close()
-	}
 	return nil
 }
 
 // Get a piece reader that is shared between callers. These readers are most
 // performant for random acccess (eg bitswap reads).
 // If there is no error, the caller must call Close() on the section reader.
 func (ps *PieceDirectory) GetSharedPieceReader(ctx context.Context, pieceCid cid.Cid) (types.SectionReader, error) {
+	ctx, span := tracing.Tracer.Start(ctx, "pm.get_shared_piece_reader")
+	defer span.End()
+	span.SetAttributes(attribute.String("piececid", pieceCid.String()))
+
+	var r *cachedSectionReader
+
 	// Check if there is already a piece reader in the cache
-	ps.prCacheLk.Lock()
-	r, haveCachedReader := ps.pieceReaderCache.Get(pieceCid)
-	if haveCachedReader {
-		// There is already a cached piece reader, just increment its reference count
-		r.refs++
-	} else {
+	rr, err := ps.pieceReaderCache.Get(pieceCid.String())
+	if err != nil {
 		// There is not yet a cached piece reader, create a new one and add it
 		// to the cache
 		r = &cachedSectionReader{
 			ps:       ps,
 			pieceCid: pieceCid,
 			ready:    make(chan struct{}),
-			refs:     1,
 		}
-		ps.pieceReaderCache.Add(pieceCid, r)
-	}
-	ps.prCacheLk.Unlock()
+		ps.pieceReaderCache.Set(pieceCid.String(), r)
```

Comment:
I think we need to do the Get then Set under a lock

Reply:
Reverted the lock.